### DISCUSSION TITLE
About the calculation of Frame velocity

### DISCUSSION THREAD

**Discussion by XuSen-Proto** [raisimTech/raisimLib#269] [Q&A] [answered]
I have implemented a bipe robot simulation using MATLAB and Raisim. I gave the states **_q_** and their derivetives **_u_**, and then chose one frame to get the Frame Jocabian _**J**_ and the frame velocity **_v_** using the API. It turns out that the equation **_v_** = **_Ju_** doesn't hold, I wonder why.

**Comment by jhwangbo** [comment, marked as answer]
can you share your code?

**Comment by XuSen-Proto** [comment]
The code is in the attachment and the execution environment is MATLAB R2021b. Run demo.m, and you will find that the frame jacobian will be OK once the function&nbsp;setGeneralizedCoordinate is called, but the Frame Velocity will not equal Ju until the functions setGeneralizedCoordinate and&nbsp;getFrameAngularVelocity are called twice. Also, the nonlinear force terms from the two calculations are different, and the result is shown in the picture. The amended MATLAB interface,&nbsp;raisim_interface_mex.cpp, where the bug of getNonlinearForce is fixed, is also given in the attachment. I'm looking forward to your reply.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "raisimTech/raisimLib"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年2月16日(星期三) 晚上10:34
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [raisimTech/raisimLib] About the calculation of Frame velocity (Discussion #269)





 
can you share your code?
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by jhwangbo** [reply]
I don't see your attachment here. Did you forget it?

**Reply by XuSen-Proto** [reply]
[DemoForDiscussion.zip](https://github.com/raisimTech/raisimLib/files/8084984/DemoForDiscussion.zip)

**Reply by jhwangbo** [reply]
The issue is resolved. You can pull and test. The issue was that "setGeneralizedVelocity" did not update the kinematics. Only "setGeneralizedVelocity" did. But now both should update the kinematics

**Reply by XuSen-Proto** [reply]
Thanks for your updating, and I have made the kinematics updated correctly in Win10 too. However, I found another problem about inverse dynamics recently. I imported iwwa14, a fixed-base manipulator, and set the damping zero, to calculate the NonlinearForce, However, the result turned out different from that solved by robotics system toolbox of MATLAB, which confused me. The code is in the attachment.  I'm looking forward to your reply.
[InverseDynamicsTest.zip](https://github.com/raisimTech/raisimLib/files/8130692/InverseDynamicsTest.zip)

**Reply by jhwangbo** [reply]
Did you match the frame? the angular velocity of a floating base in RaiSim is in the world frame

**Reply by XuSen-Proto** [reply]
Yes. Besides, it's a fixed-base system.

**Reply by jhwangbo** [reply]
Do you use the PD controller? if so, can you set the gains to zero and test again?

**Reply by XuSen-Proto** [reply]
No, I just set the gc, gv, and then call getNonlinearities.

**Reply by XuSen-Proto** [reply]
I have zeroed the gains and did another trail, and the problem still existed.

**Reply by jhwangbo** [reply]
Please post different questions in a seperate thread
